distribution/xfer/transfer.go

```diff
@@ -307,6 +336,85 @@ func (tm *transferManager) SetConcurrency(concurrency int) {
 	tm.mu.Unlock()
 }
 
+var cacheSubdir = "resumable_downloads"
+
+// hashKey hashes the key to create the directory name because otherwise tche path
+// would be too long. This hash needs to be secure or else multiple downloads could
+// share the same cache directory and cause unexpected and potentially malicious behaviour.
+func hashKey(key string) string {
+	hash := sha256.New()
+	hash.Write([]byte(key))
+	encoded := make([]byte, hash.Size()*2)
+	hex.Encode(encoded, hash.Sum(nil))
+	return string(encoded)
+}
+
+func (tm *transferManager) cachePathForKeyHash(keyHash string) string {
+	return filepath.Join(tm.cacheRoot, keyHash)
+}
+
+func (tm *transferManager) cleanUpByKeyHash(keyHash string) error {
+	return os.RemoveAll(tm.cachePathForKeyHash(keyHash))
+}
+
+// CollectGarbage purges the cache directory of any caches for transfers
+// that are not in progress
+func (tm *transferManager) CollectGarbage() error {
```

Comment:
I think if I put it in the download manager I'll have to thread it through more places, but it should still work. I also wasn't sure if we want to also have a cache for uploads, in which case this would be useful. The reason I have two caches now is that the transfer manager doesn't know about images. I'll try moving cache management entirely into the download manager. I might make a separate package/struct for managing and persisting a cache directory.